src/network/networkd-link.c

```diff
@@ -1864,45 +1844,47 @@ static int link_add(Manager *m, sd_netlink_message *message, Link **ret) {
 
         log_link_debug(link, "Link %d added", link->ifindex);
 
-        if (path_is_read_only_fs("/sys") <= 0) {
-                /* udev should be around */
-                sprintf(ifindex_str, "n%d", link->ifindex);
-                r = sd_device_new_from_device_id(&device, ifindex_str);
-                if (r < 0) {
-                        log_link_debug_errno(link, r, "Could not find device, waiting for device initialization: %m");
-                        return 0;
-                }
-
-                r = sd_device_get_is_initialized(device);
-                if (r < 0) {
-                        log_link_warning_errno(link, r, "Could not determine whether the device is initialized: %m");
-                        goto failed;
-                }
-                if (r == 0) {
-                        /* not yet ready */
-                        log_link_debug(link, "link pending udev initialization...");
-                        return 0;
-                }
-
-                r = device_is_renaming(device);
-                if (r < 0) {
-                        log_link_warning_errno(link, r, "Failed to determine the device is being renamed: %m");
-                        goto failed;
-                }
-                if (r > 0) {
-                        log_link_debug(link, "Interface is being renamed, pending initialization.");
-                        return 0;
-                }
-
-                r = link_initialized(link, device);
-                if (r < 0)
-                        goto failed;
-        } else {
+        if (path_is_read_only_fs("/sys") > 0) {
+                /* no udev */
                 r = link_initialized_and_synced(link);
                 if (r < 0)
                         goto failed;
+                return 0;
         }
 
+        /* udev should be around */
+        xsprintf(ifindex_str, "n%d", link->ifindex);
+        r = sd_device_new_from_device_id(&device, ifindex_str);
```

Comment:
given we have sd_device_new_from_devnum() and sd_device_new_from_stat_rdev() as friendly ways to get a device object for device nodes, maybe we should actually add a similar helper call: `sd_device_new_from_ifindex()` (and maybe even `sd_device_new_from_ifname()`). network interfaces after all treated in special ways all across the udev codebase, hence probably should carry their own apis in the sd-device interface too.

(certainly something for a different PR, another time)